web/app/dao/SearchDAO.java

```diff
@@ -174,10 +173,148 @@
+			Collections.sort(cachedAutoCompleteListForMetric);
+			Cache.set(SEARCH_AUTOCOMPLETE_LIST_METRIC, cachedAutoCompleteListForMetric, 60*60);
+		}
+
+		return cachedAutoCompleteListForMetric;
+	}
+
+	public static List<String> getAutoCompleteListForFlow()
+	{
+		List<String> cachedAutoCompleteListForFlow = (List<String>)Cache.get(SEARCH_AUTOCOMPLETE_LIST_FLOW);
+		if (cachedAutoCompleteListForFlow == null || cachedAutoCompleteListForFlow.size() == 0)
+		{
+			List<String> flowList = getJdbcTemplate().queryForList(GET_FLOW_AUTO_COMPLETE_LIST, String.class);
+			List<String> jobList = getJdbcTemplate().queryForList(GET_JOB_AUTO_COMPLETE_LIST, String.class);
+			cachedAutoCompleteListForFlow =
+							Stream.concat(flowList.stream(), jobList.stream()).collect(Collectors.toList());
+			Collections.sort(cachedAutoCompleteListForFlow);
+			Cache.set(SEARCH_AUTOCOMPLETE_LIST_FLOW, cachedAutoCompleteListForFlow, 60*60);
+		}
+
+		return cachedAutoCompleteListForFlow;
+	}
+
+	public static List<String> getSuggestionList(String category, String searchKeyword)
+	{
+		List<String> SuggestionList = new ArrayList<String>();
+		String elasticSearchType = "dataset";
+		String elasticSearchTypeURLKey = "elasticsearch.dataset.url";
+		String fieldName = "name";
+
+		JsonNode responseNode = null;
+		ObjectNode keywordNode = null;
+
+		try
+		{
+			String lCategory = category.toLowerCase();
+			Logger.info("lCategory is " + category);
+
+			switch (lCategory) {
+				case "dataset":
+					elasticSearchType = "dataset";
+					elasticSearchTypeURLKey = "elasticsearch.dataset.url";
+					fieldName = "name";
+					break;
+				case "metric":
+					elasticSearchType = "metric";
+					elasticSearchTypeURLKey = "elasticsearch.metric.url";
+					fieldName = "metric_name";
+					break;
+				case "flow":
+					elasticSearchType = "flow";
+					elasticSearchTypeURLKey = "elasticsearch.flow.url";
+					fieldName = "flow_name";
+					break;
+				default:
+					break;
+			}
+
+			keywordNode = utils.Search.generateElasticSearchPhraseSuggesterQuery(elasticSearchType, fieldName, searchKeyword);
+		}
+		catch(Exception e)
+		{
+			Logger.error("Elastic search phrase suggester error. Error message :" + e.getMessage());
+		}
+
+		Logger.info("The suggest query sent to Elastic Search is: " + keywordNode.toString());
+
+		Promise<WSResponse> responsePromise = WS.url(Play.application().configuration().getString(
+				elasticSearchTypeURLKey)).post(keywordNode);
+		responseNode = responsePromise.get(1000).asJson();
+
+		// Logger.info("responseNode for getSuggestionList is " + responseNode.toString());
+
+		if (responseNode != null && responseNode.isContainerNode() && responseNode.has("hits"))
+		{
+			JsonNode suggestNode = responseNode.get("suggest");
+			Logger.info("suggestNode is " + suggestNode.toString());
+			if (suggestNode != null && suggestNode.has("simple_phrase"))
+			{
+				JsonNode simplePhraseNode = suggestNode.get("simple_phrase");
```

Comment:
This [arrowhead code](https://lostechies.com/chrismissal/2009/05/27/anti-patterns-and-worst-practices-the-arrowhead-anti-pattern/) might not be the best code style. We can leave it for now but try to avoid this in the future.